README.md
Python Client for Google Maps Services
====================================

[![Build Status](https://travis-ci.org/googlemaps/google-maps-services-python.svg?branch=master)](https://travis-ci.org/googlemaps/google-maps-services-python)
[![codecov](https://codecov.io/gh/googlemaps/google-maps-services-python/branch/master/graph/badge.svg)](https://codecov.io/gh/googlemaps/google-maps-services-python)
[![PyPI version](https://badge.fury.io/py/googlemaps.svg)](https://badge.fury.io/py/googlemaps)
![PyPI - Downloads](https://img.shields.io/pypi/dd/googlemaps)
![GitHub contributors](https://img.shields.io/github/contributors/googlemaps/google-maps-services-python)

## Description

Use Python? Want to geocode something? Looking for directions?
Maybe matrices of directions? This library brings the Google Maps Platform Web
Services to your Python application.
![Analytics](https://maps-ga-beacon.appspot.com/UA-12846745-20/google-maps-services-python/readme?pixel)

The Python Client for Google Maps Services is a Python Client library for the following Google Maps
APIs:

 - Directions API
 - Distance Matrix API
 - Elevation API
 - Geocoding API
 - Geolocation API
 - Time Zone API
 - Roads API
 - Places API
 - Maps Static API

Keep in mind that the same [terms and conditions](https://developers.google.com/maps/terms) apply
to usage of the APIs when they're accessed through this library.

## Support

This library is community supported. We're comfortable enough with the stability and features of
the library that we want you to build real production applications on it. We will try to support,
through Stack Overflow, the public and protected surface of the library and maintain backwards
compatibility in the future; however, while the library is in version 0.x, we reserve the right
to make backwards-incompatible changes. If we do remove some functionality (typically because
better functionality exists or if the feature proved infeasible), our intention is to deprecate
and give developers a year to update their code.

If you find a bug, or have a feature suggestion, please log an issue. If you'd like to
contribute, please read contribute.

## Requirements

 - Python 3.5 or later.
 - A Google Maps API key.

## API Keys

Each Google Maps Web Service request requires an API key or client ID. API keys
are generated in the 'Credentials' page of the 'APIs & Services' tab of [Google Cloud console](https://console.cloud.google.com/apis/credentials).

For even more information on getting started with Google Maps Platform and generating/restricting an API key, see [Get Started with Google Maps Platform](https://developers.google.com/maps/gmp-get-started) in our docs.

**Important:** This key should be kept secret on your server.

## Installation

    $ pip install -U googlemaps

Note that you will need requests 2.4.0 or higher if you want to specify connect/read timeouts.

## Usage

This example uses the Geocoding API and the Directions API with an API key:

```python
import googlemaps
from datetime import datetime

gmaps = googlemaps.Client(key='Add Your Key here')

# Geocoding an address
geocode_result = gmaps.geocode('1600 Amphitheatre Parkway, Mountain View, CA')

# Look up an address with reverse geocoding
reverse_geocode_result = gmaps.reverse_geocode((40.714224, -73.961452))

# Request directions via public transit
now = datetime.now()
directions_result = gmaps.directions("Sydney Town Hall",
                                     "Parramatta, NSW",
                                     mode="transit",
                                     departure_time=now)
```

For more usage examples, check out [the tests](https://github.com/googlemaps/google-maps-services-python/tree/master/googlemaps/test).

## Features

### Retry on Failure

Automatically retry when intermittent failures occur. That is, when any of the retriable 5xx errors
are returned from the API.


## Building the Project


    # Installing nox
    $ pip install nox

    # Running tests
    $ nox

    # Generating documentation
    $ nox -e docs

    # Copy docs to gh-pages
    $ nox -e docs && mv docs/_build/html generated_docs && git clean -Xdi && git checkout gh-pages

## Documentation & resources
### Getting started
- [Get Started with Google Maps Platform](https://developers.google.com/maps/gmp-get-started)
- [Generating/restricting an API key](https://developers.google.com/maps/gmp-get-started#api-key)
- [Authenticating with a client ID](https://developers.google.com/maps/documentation/directions/get-api-key#client-id)

### API docs
- [Google Maps Platform web services](https://developers.google.com/maps/apis-by-platform#web_service_apis)
- [Directions API](https://developers.google.com/maps/documentation/directions/)
- [Distance Matrix API](https://developers.google.com/maps/documentation/distancematrix/)
- [Elevation API](https://developers.google.com/maps/documentation/elevation/)
- [Geocoding API](https://developers.google.com/maps/documentation/geocoding/)
- [Geolocation API](https://developers.google.com/maps/documentation/geolocation/)
- [Time Zone API](https://developers.google.com/maps/documentation/timezone/)
- [Roads API](https://developers.google.com/maps/documentation/roads/)
- [Places API](https://developers.google.com/places/)
- [Maps Static API](https://developers.google.com/maps/documentation/maps-static/)

### Support
- [Report an issue](https://github.com/googlemaps/google-maps-services-python/issues)
- [Contribute](https://github.com/googlemaps/google-maps-services-python/blob/master/CONTRIB.md)
- [StackOverflow](http://stackoverflow.com/questions/tagged/google-maps)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [v4.2.0]
### Added
- Add support for Maps Static API (#344)

## [v4.1.0]
### Added
- Adding support for passing in `experience_id` to Client class (#338)

## [v4.0.0]
### Changed
- Python 2 is no longer supported
- Removed place fields: `alt_id`, `id`, `reference`, and `scope`. Read more about this at https://developers.google.com/maps/deprecations.

## [v3.1.4]
### Changed
- `APIError.__str__` should always return a str (#328)

## [v3.1.3]
### Changed
- deprecation warning for place fields: `alt_id`, `id`, `reference`, and `scope`. Read more about this at https://developers.google.com/maps/deprecations.

## [v3.1.2]
### Added
- Tests for distribution tar as part of CI
- Support for subfields such as `geometry/location` and `geometry/viewport` in Places.

## [v3.1.1]
### Changed
- Added changelog to manifest

## [v3.1.0]
### Changed
- Switched build system to use [nox](https://nox.thea.codes/en/stable/), pytest, and codecov. Added Python 3.7 to test framework.
- Set precision of truncated latitude and longitude floats [to 8 decimals](https://github.com/googlemaps/google-maps-services-python/pull/301) instead of 6.
- Minimum version of requests increased.
- Session token parameter [added](https://github.com/googlemaps/google-maps-services-python/pull/244) to `place()`.
- Fixed issue where headers in `request_kwargs` were being overridden.
### Added
- Automation for PyPi uploads.
- Long description to package.
- Added tests to manifest and tarball.
### Removed
- Removed places `places_autocomplete_session_token` which can be replaced with `uuid.uuid4().hex`.
- Removed deprecated `places_radar`.


**Note:** Start of changelog is 2019-08-27, [v3.0.2].

[Unreleased]: https://github.com/googlemaps/google-maps-services-python/compare/4.2.0...HEAD
[v4.2.0]: https://github.com/googlemaps/google-maps-services-python/compare/4.1.0...4.2.0
[v4.1.0]: https://github.com/googlemaps/google-maps-services-python/compare/4.0.0...4.1.0
[v4.0.0]: https://github.com/googlemaps/google-maps-services-python/compare/3.1.4...4.0.0
[v3.1.4]: https://github.com/googlemaps/google-maps-services-python/compare/3.1.3...3.1.4
[v3.1.3]: https://github.com/googlemaps/google-maps-services-python/compare/3.1.2...3.1.3
[v3.1.2]: https://github.com/googlemaps/google-maps-services-python/compare/3.1.1...3.1.2
[v3.1.1]: https://github.com/googlemaps/google-maps-services-python/compare/3.1.0...3.1.1
[v3.1.0]: https://github.com/googlemaps/google-maps-services-python/compare/3.0.2...3.1.0
[v3.0.2]: https://github.com/googlemaps/google-maps-services-python/compare/3.0.1...3.0.2
[v3.0.1]: https://github.com/googlemaps/google-maps-services-python/compare/3.0.0...3.0.1
[v3.0.0]: https://github.com/googlemaps/google-maps-services-python/compare/2.5.1...3.0.0


noxfile.py
import nox

SUPPORTED_PY_VERSIONS = ["3.5", "3.6", "3.7", "3.8"]


def _install_dev_packages(session):
    session.install("-e", ".")


def _install_test_dependencies(session):
    session.install("-r", "test_requirements.txt")


def _install_doc_dependencies(session):
    session.install("sphinx")


@nox.session(python=SUPPORTED_PY_VERSIONS)
def tests(session):
    _install_dev_packages(session)
    _install_test_dependencies(session)

    session.install("pytest")
    session.run("pytest")

    session.notify("cover")


@nox.session
def cover(session):
    """Coverage analysis."""
    session.install("coverage")
    session.install("codecov")
    session.run("coverage", "report", "--show-missing")
    session.run("codecov")
    session.run("coverage", "erase")


@nox.session(python="2.7")
def docs(session):
    _install_dev_packages(session)
    _install_doc_dependencies(session)

    session.run("rm", "-rf", "docs/_build", external=True)

    sphinx_args = [
        "-a",
        "-E",
        "-b",
        "html",
        "-d",
        "docs/_build/doctrees",
        "docs",
        "docs/_build/html",
    ]

    sphinx_cmd = "sphinx-build"

    session.run(sphinx_cmd, *sphinx_args)


@nox.session()
def distribution(session):
    session.run("bash", ".travis/distribution.sh", external=True)
    session.run("python", "-c", "import googlemaps")
